opensearch/sycamore-opensearch.sh

```diff
@@ -411,6 +375,123 @@ END
+            echo "Waiting for ${name} to deploy... ${i}/${max_reps}"
+            sleep 1
+        # Case 2: COMPLETED. Task is completed, so exit
+        elif [[ "${status}" == 'COMPLETED' ]]; then
+            echo "Deployed ${name} successfully"
+            MODEL_ID="${model_id}"
+            return 0
+        # Case 3: FAILED. Handle Memory Circuit Breaker error by waiting a bit but otherwise fail out
+        elif [[ "${status}" == 'FAILED' ]]; then
+            # handle error
+            debug "Deploy task failed for ${name}"
+            local error="$(jq -r '.error' "${deploy_status_file}")"
+            debug "${error}"
+            local worker_node="$(jq -r '.worker_node[0]' "${deploy_status_file}")"
+            local error_message="$(echo "${error}" | jq -r ".${worker_node}")"
+            debug "${error_message}"
+            echo "Deploy task failed for ${name}: ${error_message}"
+            # Memory Circuit Breaker error: wait 20 seconds and the try to deploy again
+            if [[ "${error_message}" = "Memory Circuit Breaker is open, please check your resources!" ]]; then
+                local wait_time=20
+                echo "Waiting for ${wait_time}s while GC runs before retrying deploy"
+                for j in $(seq "${wait_time}"); do
+                    echo -n "."
+                    sleep 1
+                done
+                echo ""
+                spawn_deploy_model_task "${name}" "${model_id}"
+                deploy_task_id="$(jq -r '.task_id' "${deploy_model_log_file}")"
+            # Duplicate Deploy Task error: set task id I'm watching to the RUNNING task
+            # - if no running task, then either the model is deployed or we try again
+            elif [[ "${error_message}" = "Duplicate deploy model task" ]]; then
+                _curl_json -XPOST "${BASE_URL}/_plugins/_ml/tasks/_search" \
+                    -o "${deploy_task_search_file}" \
+                    --data @- <<END || die "Error searching for running deploy task"
+{
+    "query": {
+        "bool": {
+            "must": [
+                {"term": {"model_id": "${model_id}"}},
+                {
+                    "bool": {
+                        "should": [
+                            {"term": {"state": "RUNNING"}},
+                            {"match_phrase": {"error": "Memory Circuit Breaker is open, please check your resources!"))
+                        ]
+                    }
+                }
+            ]
+        }
+    },
+    "sort": [{"create_time": {"order": "DESC"}}]
+}
+END
+                if [[ $(jq -r '.hits.total.value' "${deploy_task_search_file}") ]]; then
+                    debug "No running deploy task for ${model_id}. => check whether it's deployed"
+                    model_is_deployed "${model_id}" && return 0
+                    echo "${model_id} failed to deploy. Try again after 1 sec"
+                    sleep 1
```

Comment:
What is this sleep doing?  I would think it should be after the spawn if anything, but right now the system is just stuck so you're waiting when no progress can be made.

Reply:
otherwise I can spin creating deploy tasks that fail due to duplication with 0 delay

Reply:
but yeah, bump to post-spawn